### PR TITLE
Remove spawn info when unloading npc files

### PR DESCRIPTION
**Pull request by Singe-Horizontal**
* **Addressed Issue(s)**: 
#8685 

* **Server Mode**: 
Both

* **Description of Pull Request**: 
This keeps track of the source of permanent spawn generations by adding the filenames to `spawn_data` struct.
It allows removing units spawned by these files and their associated spawn data.

### CODE REVIEW

File: src/map/mob.cpp [hunk 1]
```diff
@@ -3427,6 +3427,61 @@ void mob_add_spawn(uint16 mob_id, const struct spawn_info& new_spawn)
...
+				auto itSameMap = std::find_if(removedspawns.begin(), removedspawns.end(),
+					[&md](const spawn_info& s) { return (s.mapindex == map_id2index(md->bl.m)); });
+				if (itSameMap != removedspawns.end())
+					itSameMap->qty++; //add the found monster being deleted
+				else
+					removedspawns.push_back(spawn_info{map_id2index(md->bl.m),1}); //else create a new spawn_info
+				unit_free(bl, CLR_OUTSIGHT);
+			}
+		}
+	}
```
Comment (aleos89) [inline src/map/mob.cpp:3449]:
```suggestion
	for (block_list* bl = (block_list*)mapit_first(iter); mapit_exists(iter); bl = (block_list*)mapit_next(iter)) {
		auto* md = reinterpret_cast<mob_data*>(bl);

		if (md != nullptr && md->spawn && !strcmp(md->spawn->filepath, path)) {
			auto& removedspawns = removed_mob_spawn_data[md->db->id];
			auto itSameMap = std::find_if(removedspawns.begin(), removedspawns.end(),
				[&md](const spawn_info& s) { return (s.mapindex == map_id2index(md->bl.m)); });

			if (itSameMap != removedspawns.end())
				itSameMap->qty++; //add the found monster being deleted
			else
				removedspawns.push_back(spawn_info{map_id2index(md->bl.m),1}); //else create a new spawn_info
			unit_free(bl, CLR_OUTSIGHT);
		}
	}
```

File: src/map/mob.cpp [hunk 2]
```diff
@@ -3427,6 +3427,61 @@ void mob_add_spawn(uint16 mob_id, const struct spawn_info& new_spawn)
...
+	mapit_free(iter);
+
+	//remove wanted spawn from mob_spawn_data
+	auto remove_spawn_info = [](int mobid, const auto& rs) {
+		auto& spawns = mob_spawn_data[mobid];
```
Comment (aleos89) [inline src/map/mob.cpp:3454]:
```suggestion
		auto& spawns = mob_spawn_data[mobid];

```

File: src/map/mob.cpp [hunk 3]
```diff
@@ -3427,6 +3427,61 @@ void mob_add_spawn(uint16 mob_id, const struct spawn_info& new_spawn)
...
+	for (auto& [mobid, removedspawns] : removed_mob_spawn_data) {
+		for (const auto& rs : removedspawns) {
+			remove_spawn_info(mobid,rs);
+		}
+	}
+
+	//dynamic mobs cleaning
+	if (battle_config.dynamic_mobs) {
+		for (int i = 0; i < map_num; i++) {
+			map_data* mapdata = map_getmapdata(i);
```
Comment (aleos89) [inline src/map/mob.cpp:3474]:
```suggestion
			map_data* mapdata = map_getmapdata(i);

```

File: src/map/mob.cpp [hunk 4]
```diff
@@ -3427,6 +3427,61 @@ void mob_add_spawn(uint16 mob_id, const struct spawn_info& new_spawn)
...
+			for (int16 j = 0; j < MAX_MOB_LIST_PER_MAP; j++) {
+				spawn_data* mob = mapdata->moblist[j];
```
Comment (aleos89) [inline src/map/mob.cpp:3476]:
```suggestion
				spawn_data* mob = mapdata->moblist[j];

```

File: src/map/npc.cpp [hunk 5]
```diff
@@ -6132,6 +6139,82 @@ bool npc_unloadfile( const char* path ) {
 	return found;
 }
 
+bool npc_remove_mob_spawns(const char* path) {
+
+	int spawn_count{};
+	int unit_count{};
```
Comment (aleos89) [inline src/map/npc.cpp:6145]:
```suggestion
	int32 spawn_count = {};
	int32 unit_count = {};
```

File: src/map/npc.cpp [hunk 6]
```diff
@@ -6132,6 +6139,82 @@ bool npc_unloadfile( const char* path ) {
...
+	}
+
+	// Sort spawns by spawn quantity
+	for( auto& pair : mob_spawn_data ){
+		std::sort( pair.second.begin(), pair.second.end(), []( const spawn_info& a, const spawn_info& b ) -> bool{
+			return a.qty > b.qty;
+		} );
+	}
+
+	if(spawn_count || unit_count)
```
Comment (aleos89) [inline src/map/npc.cpp:6212]:
```suggestion
	if(spawn_count > 0 || unit_count > 0)
```

**Review by aleos89 — APPROVED** [review]
Just small nitpicks, rest looks great!

File: src/map/npc.cpp [hunk 7]
```diff
@@ -6132,6 +6139,82 @@ bool npc_unloadfile( const char* path ) {
 	return found;
 }
 
+bool npc_remove_mob_spawns(const char* path) {
+
```
Comment (Lemongrass3110) [inline src/map/npc.cpp:6143]:
```suggestion
```

**Pull request merged**